cpp/src/column/column_device_view.cu

```diff
@@ -45,54 +44,75 @@ column_device_view::column_device_view( column_view source, ptrdiff_t h_ptr, ptr
                                       source.null_count(), source.offset()},
       _num_children{source.num_children()}
 {
-  if( count_descendants(source) > _num_children ) {
-    CUDF_FAIL("Columns with grand-children are not currently supported.");
-  }
-  if( _num_children > 0 )
+  size_type num_children = source.num_children();
+  if( num_children > 0 )
   {
-    column_device_view* h_column = reinterpret_cast<column_device_view*>(h_ptr);
-    column_device_view* d_column = reinterpret_cast<column_device_view*>(d_ptr);
-    d_children = d_column;
+    auto h_column = reinterpret_cast<column_device_view*>(h_ptr);
+    auto d_column = reinterpret_cast<column_device_view*>(d_ptr);
+    auto h_end = reinterpret_cast<int8_t*>(h_column + num_children);
+    auto d_end = reinterpret_cast<int8_t*>(d_column + num_children);
+    d_children = d_column; // set member ptr to device memory
     for( size_type idx=0; idx < _num_children; ++idx )
-    { // inplace-new each child
+    { // inplace-new each child into host memory
       column_view child = source.child(idx);
-      new(h_column) column_device_view(child);
-      h_column++;
+      new(h_column) column_device_view(child,reinterpret_cast<ptrdiff_t>(h_end),reinterpret_cast<ptrdiff_t>(d_end));
```

Comment:
Why do `h_end` and `d_end` use `ptrdiff_t` when they are not differences? Why not take `column_device_view*` or `intptr_t`?

Reply:
These types were not introduced as part of this PR.
The `ptrdiff_t` was intended to be a pointer-offset into device or host memory and a better choice than just using `void*`. I think an `intptr_t` would give the impression that the memory points to integers.

Reply:
But `ptrdiff_t` is meant to be used for storing a difference or offset between pointers, not for holding a pointer itself. Also `intptr_t` isn't for pointing to integers, it's an integer type that's guaranteed to be the right size to alias to a pointer. Why not take `column_device_view*`?

Reply:
This is really outside the scope of this PR. If you want to get this changed, you can open a new issue where it can be discussed and fixed in a separate PR perhaps.
The pointers really could be some random place in memory. That fact we stick a `column_device_view` at the beginning is internal details to the ctor. We could easily put them at the end of the buffer too.